### ISSUE TITLE
Home Assistant does not properly show passive cooling?

### ISSUE THREAD

**Issue by SeBsZ** [sHedC/python-masterthermconnect#99]
It seems that we don't show the cooling state properly. My season is shown as summer and the HP function has been set to 'Cooling', my heat pump is passively cooling as the setpoint is low enough. Yet, the heatpump shows 'Heating'. You can also see in my diagnostics that the pump and brine bumps are both running. The compressor is off, so it is clearly not heating.

I've attached my integration diagnostics here:
[config_entry-mastertherm-7da32185577a95dc0894742ebfc867fd.json (1).txt](https://github.com/sHedC/python-masterthermconnect/files/11478894/config_entry-mastertherm-7da32185577a95dc0894742ebfc867fd.json.1.txt)

Let me know if you need any more info or data to debug this.

**Comment by SeBsZ** [comment]
Thanks. It looks like the app and website are a bit dumb as well, they also show 'Mode: heating'... I think maybe D_277 (Passive cooling brine pump run) indicates if we are currently passively cooling or not.

**Comment by sHedC** [comment]
> Thanks. It looks like the app and website are a bit dumb as well, they also show 'Mode: heating'... I think maybe D_277 (Passive cooling brine pump run) indicates if we are currently passively cooling or not.

Yes I think D_276 turns on when its in cooling mode, D_277 turns on  when running the cooling pump and if D_276 is on and we are circulating.

Not sure about compressor/ fan I assume compressor would be off but maybe compressor 2 is on as passive cooling has a separate pump/ fan/ compressor to heating and hot water.

I am trying with the demo system but its completely quiet everything is off at the moment, but I did see it on in the past.  I think Active cooling has a compressor (reverse one) and passive just has a separate pump circuit.

So D_276 if 1 what happens to it if the Hot water starts heating?  does it go to 0?

**Comment by sHedC** [comment]
@SeBsZ - Can you test the following:

What are the values for D_276/7 in the following scenarios:
- When Heating Mode
- When Cooling Mode but Hot Water is heating
- When Cooling Mode but Idle
- When Cooling Mode but only Circulation Pump is running

I am having no luck with the demo systems seems they have turned them off mostly for now.

Thanks

**Comment by SeBsZ** [comment]
Cooling mode (cold requested) (circulation pump, brine pump running):
"D_276": "1",
"D_277": "1",

Cooling mode, cold requested, but hot water is heating:
"D_276": "1",
"D_277": "0",

Cooling mode, no cold requested, no pumps running:
"D_276": "1",
"D_277": "0",

Cooling mode, no cold requested, but only circulation pump is running:
"D_276": "1",
"D_277": "0",

Heating mode, no heat requested, hot water heating:
"D_276": "1",
"D_277": "0",

Heating mode, heat requested, compressor and all pumps running, season winter and mode: heating:
"D_276": "1",
"D_277": "0",

So to me, it looks like we can definitely use D_277 to indicate passive cooling is active, but it seems like D_276 only indicates whether the heatpump has passive cooling support, not whether it is currently enabled or not.

**Comment by SeBsZ** [comment]
I believe D_193 only turned to 0 in the "Cooling mode, no cold requested, no pumps running:" scenario. It is "1" in all other cases, so not sure how useful that is.

D_420 and D_422 is 0 in all cases...

**Comment by SeBsZ** [comment]
Not sure what more you need? Looks like D_277 is what should be used to show active passive cooling?

Also now I'm seeing:
![image](https://github.com/sHedC/python-masterthermconnect/assets/5457531/8dda0269-857c-4252-be6d-0e658903bdc4)

Is this just a typo? Should this be "dewpoint protection"? Currently D_196=1 (dew point) but D_11 = 0 (defrost mode).

**Comment by SeBsZ** [comment]
I also discovered that when passively cooling, the requested temperature is incorrect:
![image](https://github.com/sHedC/python-masterthermconnect/assets/5457531/38ea89ca-ce97-4a2d-80e9-6b598d3ec080)

Here the actual temperature is correct, 16.4. But, the target temperature is 15 currently, which is stored in A_312. Can we make this change somehow?

**Comment by SeBsZ** [comment]
Two things:
I see this line in sensor.en.json:

```
"dpc": "Defrost Protection"
```

But dpc is Dew Point Control, not Defrost Protection. So that's wrong and an easy fix.

Another is the cooling curve, some registers are mixed up. If you look at mastertherm.online:
![image](https://github.com/sHedC/python-masterthermconnect/assets/5457531/4867b8a9-2e64-47a8-aa7e-e164a6a80381)

And then at HomeAssistant, you can see something is mixed up:
![image](https://github.com/sHedC/python-masterthermconnect/assets/5457531/974aad3c-50c6-4cc5-b349-8d13d6e209f2)

As for testing, I can always test stuff for you.

**Comment by SeBsZ** [comment]
D_193 is currently 1 while cooling and circulation pump and brine pump running.

What do you want me to test? Not sure what you meant.

**Comment by sHedC** [comment]
> D_193 is currently 1 while cooling and circulation pump and brine pump running.
> 
> What do you want me to test? Not sure what you meant.

No that is what I expected, the test data that I found, basically if circulation pump is running but D_277 is 0 then D_193 is 1 for cooling and 0 for heating.

I am updating based on logic to test for D_277 = 1 or D_193 = 1 so if only circulation pump is running but we are cooling then it shows cooling and not heating.

**Comment by sHedC** [comment]
> Also now I'm seeing: ![image](https://user-images.githubusercontent.com/5457531/245437288-8dda0269-857c-4252-be6d-0e658903bdc4.png)
> 
> Is this just a typo? Should this be "dewpoint protection"? Currently D_196=1 (dew point) but D_11 = 0 (defrost mode).

Looks correct in this repo:
defrost_mode = D_11 (when 1) and dewp_control = D_19

> Also now I'm seeing: ![image](https://user-images.githubusercontent.com/5457531/245437288-8dda0269-857c-4252-be6d-0e658903bdc4.png)
> 
> Is this just a typo? Should this be "dewpoint protection"? Currently D_196=1 (dew point) but D_11 = 0 (defrost mode).

Its a typo in the translation in the Home Assistant integration I will correct that there.

**Comment by SeBsZ** [comment]
Yeah, it was in sensor.en.json.

I will try to get the circulation pump running, and then switch between heating and cooling and give you the values of D_193?

**Comment by sHedC** [comment]
> Two things: I see this line in sensor.en.json:
> 
> ```
> "dpc": "Defrost Protection"
> ```
> 
> But dpc is Dew Point Control, not Defrost Protection. So that's wrong and an easy fix.
> 
> Another is the cooling curve, some registers are mixed up. If you look at mastertherm.online: ![image](https://user-images.githubusercontent.com/5457531/245561367-4867b8a9-2e64-47a8-aa7e-e164a6a80381.png)
> 
> And then at HomeAssistant, you can see something is mixed up: ![image](https://user-images.githubusercontent.com/5457531/245561444-974aad3c-50c6-4cc5-b349-8d13d6e209f2.png)
> 
> As for testing, I can always test stuff for you.

Yep the registers were the wrong way round.  Just checking the requirements version and testing as 3.11 is now the default so probalby moving everything up versions.

Then will update the HASS Integration

**Comment by SeBsZ** [comment]
My heatpump is currently heating for DHW, so passive cooling is temporarily off. I'm getting these registers:

```
D_193 = 1
D_4 = 0
D_276 = 1
D_277 = 0
D_422 = 0
D_420 = 0
```

Might be of use.

**Comment by sHedC** [comment]
thanks basically saying you are in cooling mode but its currently heating water, should be fine.

**Comment by SeBsZ** [comment]
Thanks. Have you been able to implement A_312 or not yet?

**Comment by sHedC** [comment]
Yes in latest fix 2.2.4 but not yet put it into hass, tomorrow will do that

Also fixed the operating mode if my guess is right

**Comment by sHedC** [comment]
All updated in the 1.1.5-b0 release of the integration, let me know once you check the above :)

**Comment by SeBsZ** [comment]
Hey, two things. So I'm running the latest 1.1.5-b0. My heat pump mode when cooling is now always 'Dew Point Control':
![image](https://github.com/sHedC/python-masterthermconnect/assets/5457531/6a462057-c267-45f4-9b9f-ecb117ac185d)

I don't believe this is an issue on your end. Dew point protection is enabled and so I think it will always show this, even when I'm not hitting the dew point yet. I checked and D_196 and D_197 are both 1, so not sure what to do about this. I think we don't need to show this in the Operating Mode at all - if I want to see this I can just use the 'Dew Point Control' entity.

Secondly, I asked if you were able to implement A_312 yet but I think you missed what this was about. See my comment here: https://github.com/sHedC/python-masterthermconnect/issues/99#issuecomment-1588980906 . The requested temperature when cooling is incorrect. Let me know if you have any questions about this.

**Comment by sHedC** [comment]
> Secondly, I asked if you were able to implement A_312 yet but I think you missed what this was about. See my comment here: [#99 (comment)](https://github.com/sHedC/python-masterthermconnect/issues/99#issuecomment-1588980906) . The requested temperature when cooling is incorrect. Let me know if you have any questions about this.

I missed that one problem with having too many Item in the same issue, will look to check and fix.  Did you check the cooling curve?


> Hey, two things. So I'm running the latest 1.1.5-b0. My heat pump mode when cooling is now always 'Dew Point Control':

Didn't change this logic, DPC is checked before we get to cooling/ heating in my logic and in the app.  What I changed comes after that check.  What does the Web App say?

                "'dhw' if {0} else "
                "'pool' if {1} else "
                "'aux_heater' if not ({2} or {3}) and ({4} or {5}) else "
                "'dpc' if {6} else "
                "'cooling' if {7} or {12} or {13} else"
                "'heating' if {8} or {9} or {10} or {11} else "
                "'idle'"

Should follow the APP logic, but you did see heating instead of cooling at some points so must mean DPC is only turned on in some circumstances?

**Comment by sHedC** [comment]
Yes could always remove the DPC check.

**Comment by SeBsZ** [comment]
Yeah, DPC isn't always shown. When heating or heating dhw it doesn't show. But, i got the installer to turn on dew point protection for safety reasons and now it is always showing when passively cooling. The website and app do not show it, but then again they show heating so they're worse than our implementation. I would just hide the dpc mode, i don't think it is necessary.

**Comment by sHedC** [comment]
going to check the app, wonder whit it does not show logic came from there.

**Comment by sHedC** [comment]
No idea where it came from shall I remove it?

**Comment by SeBsZ** [comment]
Yeah

**Comment by SeBsZ** [comment]
Yeah

**Comment by sHedC** [comment]
Um what about the translation in HASS.  

Cooling with Dew Point Protection seems too long?

**Comment by sHedC** [comment]
"Cooling (DPC)"?

**Comment by sHedC** [comment]
Ok you can try with 1.1.5 Beta 1, the new operating mode is in there.

**Comment by sHedC** [comment]
Icon is a snowflake with half water drops.

**Comment by sHedC** [comment]
Fixed and published to the home assistant integration

**Comment by SeBsZ** [comment]
I think we should reopen this ticket. All summer, the operating mode showed a nice cooling mode which was great. However, now that I turned it back to winter mode and I started heating again, the mode is showing cooling instead of heating.

In `datamapread.py` you have this logic:
```
"'cooling_dpc' if {6} and ({7} or {12} or {13}) else "
"'cooling' if {7} or {12} or {13} else "
```

I believe the issue here is with `{13}` which is:
```
[bool, "D_193"],  # 13 - Requested Mode
```

This 'requested mode', or register D_193 is discussed above in this ticket. With the current logic, if we're not heating water (dhw), no heating the pool and not running an aux heater then if D_193 is true, we end up in cooling mode. This is not correct, as I believe  D_193 can also be 1 when heating.

Maybe this needs more testing?
